engine/buildkit/telemetry.go

```diff
@@ -0,0 +1,75 @@
+package buildkit
+
+import (
+	"context"
+	"strings"
+
+	"github.com/dagger/dagger/telemetry"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/trace"
+	"go.opentelemetry.io/otel/trace/embedded"
+)
+
+// buildkitTelemetryContext returns a context with a wrapped span that has a
+// TracerProvider that can process spans produced by buildkit. This works,
+// because of how buildkit heavily relies on trace.SpanFromContext.
+func buildkitTelemetryContext(ctx context.Context) context.Context {
+	if ctx == nil {
+		return nil
+	}
+	sp := trace.SpanFromContext(ctx)
+	sp = buildkitSpan{Span: sp, provider: &buildkitTraceProvider{tp: sp.TracerProvider()}}
+	return trace.ContextWithSpan(ctx, sp)
+}
+
+type buildkitTraceProvider struct {
+	embedded.TracerProvider
+	tp trace.TracerProvider
+}
+
+func (tp *buildkitTraceProvider) Tracer(name string, options ...trace.TracerOption) trace.Tracer {
+	return &buildkitTracer{
+		tracer:   tp.tp.Tracer(name, options...),
+		provider: tp,
+	}
+}
+
+type buildkitTracer struct {
+	embedded.Tracer
+	tracer   trace.Tracer
+	provider *buildkitTraceProvider
+}
+
+func (t *buildkitTracer) Start(ctx context.Context, spanName string, opts ...trace.SpanStartOption) (context.Context, trace.Span) {
+	internal := false
+
+	if strings.HasPrefix(spanName, "load cache: ") {
+		internal = true
+	}
+
```

Comment:
Not sure about this one - I think these are still useful to show in the UI, if anything I'd rather they just didn't have the `load cache: ` prefix and instead had a `cached: true` attribute or something so we could add a proper UI cue.

This is one area I was thinking we might want to just make upstream PRs to Buildkit, since I'm not sure how opinionated they are about it or if this is just their initial rough draft of Otel integration. Or maybe we just do all this at the SpanProcessor layer; there's probably a limit to how tightly we want couple Buildkit to our UI.

I'd say for now we should just leave it alone, since we'd want to coordinate any new `cached: true` attribute with making the UI/TUI actually respect it.

Reply:
Yup, sounds good to me, have removed this case.